enforce the interface relationship between ServicePort and BaseServiceInfo

JacobTanenbaum opened the pull request:
Currently the BaseServiceInfo struct implements the ServicePort interface, but
only uses that interface sometimes. All the elements of BaseServiceInfo are exported
and sometimes the interface is used to access them and othertimes not

I extended the ServicePort interface so that all relevent values can be accessed through
it and unexported all the elements of BaseServiceInfo

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
>
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

danwinship commented on pkg/proxy/types.go:
```diff
@@ -48,8 +49,14 @@ func (spn ServicePortName) String() string {
-	// ClusterIPString returns service cluster IP in string format.
-	ClusterIPString() string
+	// GetClusterIP returns service cluster IP in net.IP format.
+	GetClusterIP() net.IP
```
golang style is not to use "Get" in getter names (https://golang.org/doc/effective_go.html#Getters)

JacobTanenbaum replied:
I used "Get" in the getter name because some of them already did and now a majority of them do now. Should I remove "Get" from all of them?

I changed from string to net.IP because sometimes the string version of the IP address and other times they just accessed the net.IP through BaseServiceInfo.ClusterIP. I thought instead of having two getters since net.IP has a string method it would be better to return the more generic type and call the string function when needed

danwinship replied:
Hm... it appears that the interface used to have `ClusterIP()`, `Protocol()`, and `HealthCheckNodePort()`, and @MrHohn changed it in https://github.com/kubernetes/kubernetes/commit/b485f7b5 as part of #56880. (Renaming the methods to have `Get` so as to not conflict with newly-added public struct fields named `ClusterIP`, `Protocol`, and `HealthCheckNodePort`.) I'm not sure if there's any good reason for it...

Anyway, if you're making the struct fields private and consistently using the interface methods, then you should make the interface method names not have "Get" in them.

MrHohn replied:
Thanks for adding me. @danwinship is correct, I made these this way purely to avoid conflicting. +1 to not having the redundant `Get` if no longer needed.

danwinship reviewed (commented):
What inspired this? Is this a prologue to some other change?

JacobTanenbaum commented:
@danwinship I found it confusing the last time I changed kubeproxy that the interface existed but wasn't used consistently

danwinship commented:
Drop pkg/scheduler/apis/config/v1alpha1/zz_generated.conversion.go from the patch. (Looks like you did an update/regen with a slightly different version of tools and got irrelevant changes.)

JacobTanenbaum committed:
enforce the interface relationship between ServicePort and BaseServiceInfo

Currently the BaseServiceInfo struct implements the ServicePort interface, but
only uses that interface sometimes. All the elements of BaseServiceInfo are exported
and sometimes the interface is used to access them and othertimes not

I extended the ServicePort interface so that all relevent values can be accessed through
it and unexported all the elements of BaseServiceInfo

JacobTanenbaum commented:
/retest

JacobTanenbaum commented:
@dcbw @danwinship PTAL

dcbw commented:
/priority important-backlog
/lgtm

dcbw commented:
@freehan @thockin you guys able to approve this one? Thanks!

JacobTanenbaum commented:
@freehan @thockin does this need more work?

JacobTanenbaum commented:
/retest

dcbw commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -993,7 +992,6 @@ func onlyLocalNodePorts(t *testing.T, fp *Proxier, ipt *iptablestest.FakeIPTable
 	fp.nodePortAddresses = []string{"10.20.30.0/24"}
 
 	fp.syncProxyRules()
-
```
unnecessary whitespace removal?

dcbw commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -276,18 +249,44 @@ func TestDeleteEndpointConnections(t *testing.T) {
+
+		fp.deleteEndpointConnections(input)
 
-		// For UDP connections, check the executed conntrack command
```
Did you mean to remove this comment?

dcbw commented:
/lgtm
/approve
